### ISSUE TITLE
Expose client-facing AF host name in M1 Provisioning Session resource

### ISSUE THREAD

**Issue by rjb1000** [5G-MAG/Standards#42]
# Context
In order to retrieve full Service Access Information from the 5GMS AF via reference point M5, the Media Session Handler in the 5GMS Client needs to be provided with the following *baseline Service Access Information parameters*:
- 5GMS AF host address (domain name or IP address).
- Provisioning Session identifier.
These parameters are combined by the Media Session Handler to form the request URL to retrieve the full Service Access Information from the 5GMS AF at reference point M5.

These two parameters correspond to the minimum baseline Service Access Information described in TS 26.501 when the 5GMS Application Provider announces the 5GMS service privately to the 5GMS-Aware Application (via reference point M8), 

There is a one-to-one correspondence in the 5GMS architecture between a Service Access Information resource exposed by the 5GMS AF at reference point M5 and a Provisioning Session resource created at reference point M1. All of the baseline Service Access Information parameters therefore need to be exposed to the 5GMS Application Provider at reference point M1 so that they can be included in a (private) M8 Service Announcement.

# Problem description
The Provisioning Session resource defined in clause 7.2.3.1 of TS 26.512 includes a `provisioningSessionId` property which is assigned by the 5GMS AF when the resource is created.

It does not, however, expose the client-facing 5GMS AF host address, which may differ from the host address used for M1 provisioning interactions for operational and/or security reasons.

At present, the client-facing 5GMS AF host address must be communicated manually to the 5GMS Application Provider, which is sub-optimal. Including this value in the `ProvisioningSession` resource would allow more flexibility to the 5GMS System operator.

# Suggested solution
Add a new array property (e.g. `applicationFunctionServerAddresses`) to the `ProvisioningSession` resource in table 7.2.3.1-1 of TS 26.512.

Note that there may be several valid Application Function hosts that can provide Service Access Information to 5GMS Clients, and they are all provided by the 5GMS AF in the M1 provisioning response. For precedent, see the Client Consumption Reporting Configuration, Dynamic Policy Invocation Configuration and Client Metrics Reporting Configuration (but not Network Assistance Configuration, curiously) in the Service Access Information resource defined in table 11.2.3.1-1 of TS 26.512.

Each member of the array is an opaque URL of the client-facing M5 API, following 5GMS URL format. In the interests of simplicity, it is proposed that the Provisioning Session ID is included as a path element so that it forms the complete URL required by the `M5_ServiceAccessInformation_retrieveServiceAccessInformation` operation specified in clause 11.2.2.

It is recommended to make this addition in Rel-16 onwards.

**Comment by rjb1000** [comment]
@ibouazizi comment: The 5GMS AF was intentionally not specified at reference point M1. The intention was that the Media Session Handler is pre-configured with the 5GMS AF host address(es), for example by the operator.

TS 26.512 could specify a well-known FQDN for the 5GMS AF that resolves in the local network to the correct IP address using DNS.

Alternatively, there could be something defined in the NRF. This would require discussion with SA2.

**Comment by rjb1000** [comment]
> @ibouazizi comment: The 5GMS AF was intentionally not specified at reference point M1. The intention was that the Media Session Handler is pre-configured with the 5GMS AF host address(es), for example by the operator.
> 
> TS 26.512 could specify a well-known FQDN for the 5GMS AF that resolves in the local network to the correct IP address using DNS.
> 
> Alternatively, there could be something defined in the NRF. This would require discussion with SA2.

It is unclear in both of @ibouazizi's alternative proposed solutions how the address of the 5GMS AF is made available to the Media Session Handler in the collaboration scenario where it is deployed outside the Trusted DN.

**Comment by ibouazizi** [comment]
@rjb1000 I agree with Richard. If the AF is hosted by the AP in the untrusted domain, then we would need a way to convey the address to the MSH, either through M8-M6 or M1-M5 or for a common solution it would be M1 - Nnrf.
I still prefer a common solution for all scenarios, so my preference would be for the last proposal. We will seek to start the dialogue with SA2 on this topic in SA4.

**Comment by rjb1000** [comment]
To develop your proposal a bit, @ibouazizi, I think you suggesting something along the following lines:
- When it creates a new Provisioning Session, the 5GMS AF also squirrels away some information in the NRF mapping some lookup key (e.g. Provisioning Session ID) to its client-facing hostname at M5.
   - (This mapping need to be tied to the life-cycle of the Provisioning Session. When the Provisioning Session is destroyed, the corresponding mapping needs to be removed from the NRF's mapping table.)
- At the start of media streaming session, the Media Session Handler first needs to interrogate this information via the User Plane in order to determine which 5GMS AF instance to contact to obtain full Service Access Information (and to later invoke the various Media Session Handling network APIs) via reference point M5.

A first-order approximation of the changes required for this outline solution appears to be:
1. **New lookup table in the NRF** mapping from 5GMS Provisioning Session ID to 5GMS AF client-facing host name.
    - Note that the Provisioning Session ID is not currently required to be globally unique, so different 5GMS AF instances (deployed inside or outside the Trusted DN) could assign duplicate values.
      - This problem could be solved by instead indexing the NRF's lookup table with a composite key _{Application ID, Provisioning Session ID}_.
1. **New reference point between 5GMS AF and NRF** to allow above mapping table to be populated by any authorised 5GMS AF.
    - Note that the network API also needs to be exposed via NEF to cater for 5GMS AF instances deployed outside the Trusted DN.
1. **New reference point between Media Session Handler and NRF** to look up the client-facing 5GMS host name to be used for a given media streaming session.
    - 5GMS-Aware Application first needs to supply to the Media Session Handler with the relevant Application ID and Provisioning Session ID values that it has previously obtained from the 5GMS Application Provider via M8.
      - This requires an extension to the M6 client API. (See, for example, 5GMS/rt-5gms-media-session-handler#8.)

This all seems quite a tall order, but it does appear to close the gap that has been identified, and would allow 5G Media Streaming to move from an unimplementable condition to an implementable one.

**Comment by tlohmar** [comment]
Hello, 

btw, there is an informative Annex B in TS 26.501, with title: "MNO-specific Service Access Information acquisition". The Annex also considers usage of GSMA managed FQDNs (i.e. 3gppnetworks.org).

**Comment by rjb1000** [comment]
> btw, there is an informative Annex B in TS 26.501, with title: "MNO-specific Service Access Information acquisition". The Annex also considers usage of GSMA managed FQDNs (i.e. 3gppnetworks.org).

Thanks for pointing this out, @tlohmar. I found this informative statement in clause B.1:

> When a 5GMSd-Aware Application is deployed in different 5G Systems the 5GMSd Client needs to acquire Service Access Information that resolves to the 5GMSd AF endpoint address(es) appropriate for the serving 5G System. The Service Access Information contains the URLs and API parameters of the configured 5GMSd AFs and ASs of that 5G Media Streaming System.

**Comment by rjb1000** [comment]
Having re-read annex B in TS 26.501, I have summarised my understanding in the list of solutions at the top.

**Comment by rjb1000** [comment]
This could be specified in Rel-18 as part of the "5GMS_Pro_Ph2" Work Item.

Is it acceptable to leave this underspecified in Rel-17 and Rel-16?

**Comment by rjb1000** [comment]
Draft specification text proposal contributed to SA4 _ad hoc_ meeting:

- TS 26.512 Rel-16 - _Out of scope._
- TS 26.512 Rel-17 - _Out of scope._
- TS 26.512 Rel-18 CR0036 in [S4aI230104](https://www.3gpp.org/ftp/TSG_SA/WG4_CODEC/3GPP_SA4_AHOC_MTGs/SA4_MBS/Docs/S4aI230104.zip). See clause 6.1A.2.

**Comment by rjb1000** [comment]
CR contributed to SA4#125 by @rjb1000:

* TS 26.512
  * Rel-18: **CR0036** "API changes to support 3GPP Service URL" in [S4-231161](https://www.3gpp.org/ftp/TSG_SA/WG4_CODEC/TSGS4_125_Gotheneburg/Docs/S4-231161.zip).

**Comment by rjb1000** [comment]
CR contributed to SA4#126 by @rjb1000 was endorsed as the basis of future agreement:

* TS 26.512
  * Rel-18: **CR0036**r3 "Rel-18 API changes to support 3GPP Service URL" in [S4-231636](https://www.3gpp.org/ftp/TSG_SA/WG4_CODEC/TSGS4_126_Chicago/Docs/S4-231636.zip).
